--- boot_atheros/utility/binfile/binfile.c	1970-01-01 07:00:00.000000000 +0700
+++ boot_liteon/utility/binfile/binfile.c	2013-03-29 17:58:49.000000000 +0800
@@ -0,0 +1,66 @@
+/*
+ * function: create file with hex number
+ * usage: binfile [file] [0xHH..HH]
+ *        [0xHH..HH]: from 2 to 8 digital
+ */
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <arpa/inet.h>
+
+int main(int argc, char **argv)
+{
+	int num, len, ret;
+	FILE *fp;
+
+	if(argc != 3)
+	{
+		printf("binfile usage: binfile <0xHH..HH>\n");
+		printf("[0xHH..HH]: from 2 to 8 digital\n");
+		return -1;
+	}
+
+	if(strstr(argv[2], "0x"))
+	{
+		if(sscanf(argv[2], "0x%x", &num) !=1 )
+		{
+			printf("binfile: input number format error\n");
+			return -1;
+		}
+
+		len = (strlen(argv[2]) - 2)/2;
+	}
+
+	fp=fopen(argv[1], "w+");
+
+	if(!fp)
+	{
+		printf("binfile: create file %s fail\n", argv[1]);
+		return -1;
+	}
+
+	printf("num=%x\n", num);
+	
+	num = htonl((unsigned int)num);
+
+	printf("htonl num=%x\n", num);
+
+	ret = fwrite(&num, len, 1, fp);
+
+	if(ret <=0 )
+	{
+		printf("bin: write to file %s fail\n", argv[1]);
+		return -1;
+	}
+
+	fclose(fp);
+
+	return 0;
+}
+
+
+
