Index: linux-2.6.31/arch/mips/atheros/setup.c
===================================================================
--- linux-2.6.31.orig/arch/mips/atheros/setup.c	2014-01-20 20:42:14.804112021 +0800
+++ linux-2.6.31/arch/mips/atheros/setup.c	2014-01-20 20:52:45.280090111 +0800
@@ -246,6 +246,8 @@
 	ath_reg_wr(ATH_RESET,
 		      (ath_reg_rd(ATH_RESET) & (~ATH_RESET_WMAC)));
 #endif
+        /*Delay 100ms for booting hang up, sirian@liteon, 2014-01-19*/
+        udelay(100 *1000);
 	serial_print("Booting %s\n", get_system_type());
 }
 
