Index: linux-2.6.31/drivers/net/ethernet/athrs_sgmii.c
===================================================================
--- linux-2.6.31.orig/drivers/net/ethernet/athrs_sgmii.c	2014-03-20 10:39:55.999430766 +0800
+++ linux-2.6.31/drivers/net/ethernet/athrs_sgmii.c	2014-03-20 14:08:11.167598155 +0800
@@ -172,6 +172,9 @@
 		status = (athr_reg_rd(SGMII_DEBUG_ADDRESS) & 0xff);
 	}
 
+        /*Decrease SGMII TXDR to reduce noise for wifi channel 112 issue, sirian@liteon, 2014-03-20----->*/ 
+        athr_reg_wr(SGMII_SERDES_ADDRESS, (athr_reg_rd(SGMII_SERDES_ADDRESS) & (~SGMII_SERDES_TX_DR_CTRL_MASK) |                SGMII_SERDES_HALF_TX_SET(1)));
+        /*<-----sirian@liteon, 2014-03-20*/
 	printk (MODULE_NAME "::%s  Done\n",__func__);
 
 }
Index: linux-2.6.31/drivers/net/ethernet/phys/athrs17_phy.c
===================================================================
--- linux-2.6.31.orig/drivers/net/ethernet/phys/athrs17_phy.c	2014-03-20 11:39:14.311477305 +0800
+++ linux-2.6.31/drivers/net/ethernet/phys/athrs17_phy.c	2014-03-20 14:13:08.303598088 +0800
@@ -700,7 +700,10 @@
 
 //liteon+            
             //athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc74164d0); /* SGMII control */
-            athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc78164de); /* SGMII control */
+            //athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc78164de); /* SGMII control */
+            /*Decrease SGMII TXDR to reduce noise for wifi channel 112 issue, sirian@liteon,2014-03-20*/
+            athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc78160fe); /* SGMII control */
+
 //liteon-            
             athrs17_reg_write(S17_P0STATUS_REG, S17_PORT_STATUS_AZ_DEFAULT);
             break;
@@ -748,7 +751,9 @@
             athrs17_reg_write(S17_P6STATUS_REG, S17_PORT_STATUS_AZ_DEFAULT);
 //liteon+            
             //athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc74164d0); /* SGMII control */
-            athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc78164de); /* SGMII control */
+            //athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc78164de); /* SGMII control */
+            /*Decrease SGMII TXDR to reduce noise for wifi channel 112 issue, sirian@liteon,2014-03-20*/
+            athrs17_reg_write(S17_SGMII_CTRL_REG , 0xc78160fe); /* SGMII control */
 //liteon-            
 
             break;
