--- linux-2.6.31_atheros/kernel/printk.c	2013-04-03 13:30:30.000000000 +0800
+++ linux-2.6.31_liteon/kernel/printk.c	2013-02-04 09:36:17.000000000 +0800
@@ -45,6 +45,8 @@
 
 #define __LOG_BUF_LEN	(1 << CONFIG_LOG_BUF_SHIFT)
 
+/* show only serious then KERN_NOTICE(not include). Peter, 2012-03-27. */
+#if 0
 /* printk's without a loglevel use this.. */
 #define DEFAULT_MESSAGE_LOGLEVEL 4 /* KERN_WARNING */
 
@@ -52,6 +54,17 @@
 #define MINIMUM_CONSOLE_LOGLEVEL 1 /* Minimum loglevel we let people use */
 #define DEFAULT_CONSOLE_LOGLEVEL 7 /* anything MORE serious than KERN_DEBUG */
 
+#else
+
+/* printk's without a loglevel use this.. */
+#define DEFAULT_MESSAGE_LOGLEVEL 5 /* KERN_NOTICE */
+
+/* We show everything that is MORE important than this.. */
+#define MINIMUM_CONSOLE_LOGLEVEL 1 /* Minimum loglevel we let people use */
+#define DEFAULT_CONSOLE_LOGLEVEL  5 /* anything MORE serious than KERN_NOTICE */
+
+#endif
+
 DECLARE_WAIT_QUEUE_HEAD(log_wait);
 
 int console_printk[4] = {
