--- u-boot-freescale-p1020-sku/drivers/mtd/spi/spi_flash.c	2012-07-11 19:41:34.395514589 -0700
+++ u-boot-liteon-wp854-openwrt/drivers/mtd/spi/spi_flash.c	2011-03-16 04:54:54.000000000 -0700
@@ -93,7 +93,7 @@
 
 	buffer = (unsigned char *)malloc(2 * cmd_len + data_len);
 	if (!buffer) {
-		debug("SF: Failed to malloc memory.\n");
+		printf("SF: Failed to malloc memory.\n");
 		return 1;
 	}
 	memcpy(buffer, cmd, cmd_len);
@@ -107,7 +107,7 @@
 	spi->transfer = &transfer[0];
 	ret = espi_xfer(spi);
 	if (ret) {
-		debug("SF: Failed to read %zu bytes of data: %d\n",
+		printf("SF: Failed to read %zu bytes of data: %d\n",
 				data_len, ret);
 	}
 
