Index: linux-2.6.36_MTK4110/drivers/mtd/ralink/ralink_spi.c
===================================================================
--- linux-2.6.36_MTK4110.orig/drivers/mtd/ralink/ralink_spi.c	2013-07-06 10:36:04.103174651 +0800
+++ linux-2.6.36_MTK4110/drivers/mtd/ralink/ralink_spi.c	2013-07-06 10:37:41.855886434 +0800
@@ -38,51 +38,147 @@
 #include <linux/spi/flash.h>
 #endif
 
+/*Added for using linux command line, sirian@liteon, 2013-07-05----->*/
+#ifdef LITEON_SRC_VALID
+static const char *part_probes[] __initdata = { "cmdlinepart", "RedBoot", NULL };	
+#endif
+/*<-----sirian@liteon,2013-07-05*/
+
+/*modify for wp777 nvram, daniel@liteon,2012-10-15----->*/
+#ifdef LITEON_SRC_VALID
+#define	BOOT_PART_SIZE			MTD_BOOT_PART_SIZE
+#define	BOOT_PART_OFFSET			0
+
+#define	CONFIG_PART_SIZE			MTD_CONFIG_PART_SIZE
+#define	CONFIG_PART_OFFSET		(BOOT_PART_OFFSET    + BOOT_PART_SIZE )
+
+#define	KERN_PART_SIZE			MTD_KERN_PART_SIZE
+#define	KERN_PART_OFFSET			(CONFIG_PART_OFFSET + CONFIG_PART_SIZE)
+
+#define	ROOTFS_PART_SIZE			MTD_ROOTFS_PART_SIZE
+#define	ROOTFS_PART_OFFSET	       (KERN_PART_OFFSET +  KERN_PART_SIZE)
+
+//LINUX_PART is a combination of KERN_PART and ROOTFS_PART
+//#define	LINUX_PART_SIZE			(KERN_PART_SIZE+ROOTFS_PART_SIZE)
+//#define	LINUX_PART_OFFSET			KERN_PART_OFFSET
+//don't care the above when write this mtd allocate size and offset 
+
+#define	NVRAM_PART_SIZE			MTD_NVRAM_PART_SIZE
+#define	NVRAM_PART_OFFSET		(ROOTFS_PART_OFFSET + ROOTFS_PART_SIZE )
+
+#define	BOARD_DATA_PART_SIZE			MTD_BOARD_DATA_PART_SIZE
+#define	BOARD_DATA_PART_OFFSET		(NVRAM_PART_OFFSET +NVRAM_PART_SIZE	 )
+
+#define	FACTORY_PART_SIZE			MTD_FACTORY_PART_SIZE
+#define	FACTORY_PART_OFFSET		(BOARD_DATA_PART_OFFSET  +BOARD_DATA_PART_SIZE )
+#endif //LITEON_SRC_VALID
+/*<-----daniel@liteon,2012-10-15*/
+
+
+
 static struct mtd_partition rt2880_partitions[] = {
+#ifdef LITEON_SRC_VALID
 	{
                 name:           "ALL",
-                size:           MTDPART_SIZ_FULL,
-                offset:         0,
+                size:             MTDPART_SIZ_FULL,
+                offset:          0,
         },
+#endif
 	/* Put your own partition definitions here */
-        {
-                name:           "Bootloader",
-                size:           MTD_BOOT_PART_SIZE,
-                offset:         0,
-        }, {
-                name:           "Config",
-                size:           MTD_CONFIG_PART_SIZE,
-                offset:         MTDPART_OFS_APPEND
+        {	
+/*modify for wp777, daniel@liteon,2012-10-15----->*/
+	#ifdef LITEON_SRC_VALID
+	        name:           "boot",
+		  size:            BOOT_PART_SIZE,
+               offset:         BOOT_PART_OFFSET,
+	#else
+	          name:           "Bootloader",
+		    size:            MTD_BOOT_PART_SIZE,
+		 offset:         0,
+	#endif //LITEON_SRC_VALID
+/*<-----daniel@liteon,2012-10-15*/				
         }, {
-                name:           "Factory",
-                size:           MTD_FACTORY_PART_SIZE,
-                offset:         MTDPART_OFS_APPEND
+                
+/*modify for wp777, daniel@liteon,2012-10-15----->*/
+#ifdef LITEON_SRC_VALID
+              name:           "boot_env",
+		 size:          CONFIG_PART_SIZE,
+              offset:          CONFIG_PART_OFFSET,
+#else
+		name:           "Config",
+		   size:           MTD_CONFIG_PART_SIZE,
+		offset:         MTDPART_OFS_APPEND
+	}, {
+		name:           "Factory",
+		size:           MTD_FACTORY_PART_SIZE,
+		offset:         MTDPART_OFS_APPEND
+#endif //LITEON_SRC_VALID
+/*<-----daniel@liteon,2012-10-15*/	
+	
+			
 #ifdef CONFIG_RT2880_ROOTFS_IN_FLASH
         }, {
-                name:           "Kernel",
-                size:           MTD_KERN_PART_SIZE,
-                offset:         MTDPART_OFS_APPEND,
+/*modify for wp777, daniel@liteon,2012-10-15----->*/
+#ifdef LITEON_SRC_VALID
+                name:           "kernel",			
+		   size:           KERN_PART_SIZE,
+                offset:          KERN_PART_OFFSET,
+#else
+		  name:           "Kernel",
+		  size:           MTD_KERN_PART_SIZE,	
+		  offset:         MTDPART_OFS_APPEND,
+#endif //LITEON_SRC_VALID
+/*<-----daniel@liteon,2012-10-15*/			
         }, {
-                name:           "RootFS",
-                size:           MTD_ROOTFS_PART_SIZE,
-                offset:         MTDPART_OFS_APPEND,
+
+/*modify for wp777, daniel@liteon,2012-10-15----->*/
+#ifdef LITEON_SRC_VALID		
+                name:           "rootfs",               
+	          size:           ROOTFS_PART_SIZE,
+                offset:         ROOTFS_PART_OFFSET,
+#else		
+		   name:           "RootFS",
+		   size:           MTD_ROOTFS_PART_SIZE,
+		   offset:         MTDPART_OFS_APPEND,
+#endif //LITEON_SRC_VALID	
+/*<-----daniel@liteon,2012-10-15*/	
 #ifdef CONFIG_ROOTFS_IN_FLASH_NO_PADDING
         }, {
                 name:           "Kernel_RootFS",
-                size:           MTD_KERN_PART_SIZE + MTD_ROOTFS_PART_SIZE,
+                size:            MTD_KERN_PART_SIZE + MTD_ROOTFS_PART_SIZE,
                 offset:         MTD_BOOT_PART_SIZE + MTD_CONFIG_PART_SIZE + MTD_FACTORY_PART_SIZE,
 #endif
+
+/*modify for wp777, daniel@liteon,2012-10-15----->*/
+#ifdef LITEON_SRC_VALID
+       } , {
+                name:           "nvram", 
+                size:            NVRAM_PART_SIZE,
+                offset:         NVRAM_PART_OFFSET,             
+	}, {
+		 name:           "board_data",
+                size:            BOARD_DATA_PART_SIZE,
+                offset:         BOARD_DATA_PART_OFFSET
+	},{
+                name:           "rf_data",
+                size:            FACTORY_PART_SIZE,
+                offset:         FACTORY_PART_OFFSET
+#endif //LITEON_SRC_VALID	
+/*<-----daniel@liteon,2012-10-15*/	
+				
 #else //CONFIG_RT2880_ROOTFS_IN_RAM
         }, {
                 name:           "Kernel",
                 size:           MTD_KERN_PART_SIZE,
                 offset:         MTDPART_OFS_APPEND,
 #endif
+
 #ifdef CONFIG_DUAL_IMAGE
         }, {
                 name:           "Kernel2",
                 size:           MTD_KERN2_PART_SIZE,
                 offset:         MTD_KERN2_PART_OFFSET,
+                
 #ifdef CONFIG_RT2880_ROOTFS_IN_FLASH
         }, {
                 name:           "RootFS2",
@@ -1374,6 +1470,14 @@
 {
 	struct chip_info		*chip;
 	unsigned			i;
+	
+/*Added for using linux command line, sirian@liteon, 2013-07-05----->*/
+#ifdef LITEON_SRC_VALID
+	int np;
+	struct mtd_partition *mtd_parts;
+#endif
+/*<-----sirian@liteon,2013-07-05*/
+
 #ifdef CONFIG_ROOTFS_IN_FLASH_NO_PADDING
 	loff_t offs;
 	struct __image_header {
@@ -1449,12 +1553,40 @@
 	}
 #endif
 
-
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
-	return add_mtd_partitions(&flash->mtd, rt2880_partitions, ARRAY_SIZE(rt2880_partitions));
-#else
-	add_mtd_partitions(&flash->mtd, rt2880_partitions, ARRAY_SIZE(rt2880_partitions));
-	return &flash->mtd;
+#if defined (LITEON_SRC_VALID) && defined (CONFIG_MTD_CMDLINE_PARTS)
+{
+	#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
+	{
+		int ret = 0;
+		np = parse_mtd_partitions(&flash->mtd, part_probes, &mtd_parts, 0);
+		if (np > 0) {
+			ret = add_mtd_partitions(&flash->mtd, mtd_parts, np);
+		} else {
+			printk("No partitions found on flash bank %d\n", i);
+		}
+		return ret;
+	}
+	#else
+	{
+		np = parse_mtd_partitions(&flash->mtd, part_probes, &mtd_parts, 0);
+		if (np > 0) {
+			add_mtd_partitions(&flash->mtd, mtd_parts, np);
+		} else {
+			printk("No partitions found on flash bank %d\n", i);
+		}
+		return &flash->mtd;
+	}
+	#endif
+}	
+#else // undef  CONFIG_MTD_CMDLINE_PARTS
+{
+	#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
+		return add_mtd_partitions(&flash->mtd, rt2880_partitions, ARRAY_SIZE(rt2880_partitions));
+	#else
+		add_mtd_partitions(&flash->mtd, rt2880_partitions, ARRAY_SIZE(rt2880_partitions));
+		return &flash->mtd;
+	#endif
+}
 #endif
 }
 
