diff -urN a/rt2860v2/common/cmm_info.c b/rt2860v2/common/cmm_info.c
--- a/rt2860v2/common/cmm_info.c	2013-05-13 18:14:33.121972681 +0800
+++ b/rt2860v2/common/cmm_info.c	2013-05-13 18:15:56.309759578 +0800
@@ -3510,7 +3510,7 @@
 	INT         Rssi = 0;
 	UINT        Rssi_Quality = 0;
 	NDIS_802_11_NETWORK_TYPE    wireless_mode;
-	CHAR		Ssid[MAX_LEN_OF_SSID +1];
+	CHAR		Ssid[MAX_LEN_OF_SSID + 3];
 	STRING		SecurityStr[32] = {0};
 	NDIS_802_11_ENCRYPTION_STATUS	ap_cipher = Ndis802_11EncryptionDisabled;
 	NDIS_802_11_AUTHENTICATION_MODE	ap_auth_mode = Ndis802_11AuthModeOpen;
@@ -3520,17 +3520,21 @@
 
 
 		/*SSID*/
-	NdisZeroMemory(Ssid, (MAX_LEN_OF_SSID +1));
-	if (RTMPCheckStrPrintAble((PCHAR)pBss->Ssid, pBss->SsidLen))
-		NdisMoveMemory(Ssid, pBss->Ssid, pBss->SsidLen);
+	NdisZeroMemory(Ssid, (MAX_LEN_OF_SSID +3));
+	Ssid[0] = '"';
+	if (RTMPCheckStrPrintAble((PCHAR)pBss->Ssid, pBss->SsidLen)) {
+		NdisMoveMemory(Ssid+1, pBss->Ssid, pBss->SsidLen);
+		Ssid[pBss->SsidLen+1]= '"';
+	}
 	else
 	{
 		INT idx = 0;
 		sprintf(Ssid, "0x");
 		for (idx = 0; (idx < 14) && (idx < pBss->SsidLen); idx++)
 			sprintf(Ssid + 2 + (idx*2), "%02X", (UCHAR)pBss->Ssid[idx]);
+		Ssid[1 + 2 + (idx*2)] = '"';
 	}
-		sprintf(msg+strlen(msg),"%-33s", Ssid);      
+		sprintf(msg+strlen(msg),"%-35s", Ssid);
 		
 		/*BSSID*/
 		sprintf(msg+strlen(msg),"%02x:%02x:%02x:%02x:%02x:%02x   ", 
@@ -3719,7 +3723,7 @@
 	memset(msg, 0 , TotalLen);
 	sprintf(msg,"%s","\n");
 
-	sprintf(msg+strlen(msg),"%-4s%-33s%-20s%-23s%-9s%-7s%-7s%-3s\n",
+	sprintf(msg+strlen(msg),"%-4s%-35s%-20s%-23s%-9s%-7s%-7s%-3s\n",
 	    "Ch", "SSID", "BSSID", "Security", "Siganl(%)", "W-Mode", " ExtCH"," NT");	
 
 #ifdef WSC_INCLUDED
