--- a/rt2860v2/include/oid.h
+++ b/rt2860v2/include/oid.h
@@ -66,7 +66,9 @@
 /*#define MAX_NUM_OF_CHS					49 */ /* 14 channels @2.4G +  12@UNII + 4 @MMAC + 11 @HiperLAN2 + 7 @Japan + 1 as NULL terminationc */
 /*#define MAX_NUM_OF_CHS             		54 */ /* 14 channels @2.4G +  12@UNII(lower/middle) + 16@HiperLAN2 + 11@UNII(upper) + 0 @Japan + 1 as NULL termination */
 #define MAX_NUMBER_OF_EVENT				10	/* entry # in EVENT table */
-#define MAX_NUMBER_OF_MAC				32	/* if MAX_MBSSID_NUM is 8, this value can't be larger than 211 */
+
+/* Modify MAX_NUMBER_OF_MAC's value from 32 to 33, for really support 32 AP clients, Sean@Liteon, 2013-11-04 */
+#define MAX_NUMBER_OF_MAC				33	/* if MAX_MBSSID_NUM is 8, this value can't be larger than 211 */
 
 #define MAX_NUMBER_OF_ACL				64
 #define MAX_LENGTH_OF_SUPPORT_RATES		12	/* 1, 2, 5.5, 11, 6, 9, 12, 18, 24, 36, 48, 54 */
