--- a/include/version.h
+++ b/include/version.h
@@ -188,7 +188,22 @@
   #endif
 #endif
 
-
+/*Add liteon private version,Sirian@liteon,2013-01-14----->*/
+#ifdef LITEON_SRC_VALID
+#define SHOW_VER_STR()	\
+	do {	\
+		printf("============================================ \n"); \
+		printf("Liteon UBoot Version: %s\n", LITEON_UBOOT_VERSION); \
+		printf("-------------------------------------------- \n"); \
+		printf("%s %s %s\n",CHIP_TYPE, CHIP_VERSION, GMAC_MODE); \
+		printf("DRAM component: %d Mbits %s\n", DRAM_COMPONENT, DDR_INFO); \
+		printf("DRAM bus: %d bit\n", DRAM_BUS); \
+		printf("Total memory: %d MBytes\n", DRAM_SIZE); \
+		printf("%s\n", FLASH_MSG); \
+		printf("%s\n", "Date:" __DATE__ "  Time:" __TIME__ ); \
+		printf("============================================ \n"); \
+	}while(0)
+#else
 #define SHOW_VER_STR()	\
 	do {	\
 		printf("============================================ \n"); \
@@ -202,5 +217,7 @@
 		printf("%s\n", "Date:" __DATE__ "  Time:" __TIME__ ); \
 		printf("============================================ \n"); \
 	}while(0)
+#endif//LITEON_SRC_VALID
+/*<-----Sirian@liteon,2013-01-14*/
 
 #endif	/* __VERSION_H__ */
