--- u-boot-freescale-p1020-sku/common/env_common.c	2009-12-15 14:20:54.000000000 -0800
+++ u-boot-liteon-wp854-openwrt/common/env_common.c	2011-11-08 23:26:37.678802000 -0800
@@ -133,6 +133,15 @@
 #if defined(CONFIG_PCI_BOOTDELAY) && (CONFIG_PCI_BOOTDELAY > 0)
 	"pcidelay="	MK_STR(CONFIG_PCI_BOOTDELAY)	"\0"
 #endif
+
+/*Added by sirian, 2011-07-26, for disable some hardware function*/
+#if defined(CONFIG_P1011) ||defined(CONFIG_P1020) || defined(CONFIG_P2020)
+        "disbalei2c=" "0" "\0"
+        "disbalespi=" "0" "\0"
+        "disbaletdm=" "0" "\0"
+        "disbalecore1=" "0" "\0"
+#endif
+
 #ifdef  CONFIG_EXTRA_ENV_SETTINGS
 	CONFIG_EXTRA_ENV_SETTINGS
 #endif
